File: packages/executors/src/scripted.ts
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, isAbsolute, join, relative, resolve } from "node:path";
import { ExecutionError, type ExecutionBackend, type ExecutionOutcome, type ExecutionRequest, type ExecutorPlugin } from "./types.js";

interface ScriptedScenario {
  writes: Record<string, string>;
  result: Record<string, unknown>;
  returnCode: number;
  delaySeconds: number;
  stderr: string;
  extraStdout: string;
}

function loadScenario(path: string): ScriptedScenario {
  let payload: unknown;
  try {
    payload = JSON.parse(readFileSync(path, "utf8"));
  } catch (error) {
    throw new ExecutionError(`could not load scripted scenario ${path}: ${error}`);
  }
  if (!payload || typeof payload !== "object" || Array.isArray(payload)) {
    throw new ExecutionError("scripted scenario must be a JSON object");
  }
  const record = payload as Record<string, unknown>;
  const writes = record.writes ?? {};
  if (
    !writes ||
    typeof writes !== "object" ||
    Array.isArray(writes) ||
    Object.entries(writes).some(([key, value]) => typeof key !== "string" || typeof value !== "string")
  ) {
    throw new ExecutionError("scripted scenario writes must be an object of string paths to strings");
  }
  const result = record.result;
  if (!result || typeof result !== "object" || Array.isArray(result)) {
    throw new ExecutionError("scripted scenario result must be an object");
  }
  const returnCode = record.returnCode ?? 0;
  if (!Number.isInteger(returnCode)) {
    throw new ExecutionError("scripted scenario returnCode must be an integer");
  }
  const delaySeconds = record.delaySeconds ?? 0;
  if (typeof delaySeconds !== "number" || delaySeconds < 0) {
    throw new ExecutionError("scripted scenario delaySeconds must be a non-negative number");
  }
  const stderr = record.stderr ?? "";
  if (typeof stderr !== "string") {
    throw new ExecutionError("scripted scenario stderr must be a string");
  }
  const extraStdout = record.extraStdout ?? "";
  if (typeof extraStdout !== "string") {
    throw new ExecutionError("scripted scenario extraStdout must be a string");
  }
  return {
    writes: writes as Record<string, string>,
    result: result as Record<string, unknown>,
    returnCode: returnCode as number,
    delaySeconds,
    stderr,
    extraStdout,
  };
}

function safeWritePath(cwd: string, rawPath: string): string {
  if (isAbsolute(rawPath) || rawPath.split("/").includes("..") || rawPath.split("\\").includes("..")) {
    throw new ExecutionError(`scripted write path escapes cwd: '${rawPath}'`);
  }
  const path = resolve(cwd, rawPath);
  const rel = relative(resolve(cwd), path);
  if (rel.startsWith("..") || isAbsolute(rel)) {
    throw new ExecutionError(`scripted write path escapes cwd: '${rawPath}'`);
  }
  return path;
}

export class ScriptedExecutionBackend implements ExecutionBackend {
  async execute(request: ExecutionRequest): Promise<ExecutionOutcome> {
    if (request.provider !== "scripted") {
      throw new ExecutionError(`backend 'scripted' cannot execute provider '${request.provider}'`);
    }
    const scenarioPath = join(request.cwd, ".durable-threads", "scripted.json");
    if (!existsSync(scenarioPath)) {
      throw new ExecutionError(`scripted scenario not found: ${scenarioPath}`);
    }
    const scenario = loadScenario(scenarioPath);
    await new Promise((resolveWait) => setTimeout(resolveWait, scenario.delaySeconds * 1000));
    for (const [rawPath, content] of Object.entries(scenario.writes)) {
      const path = safeWritePath(request.cwd, rawPath);
      mkdirSync(dirname(path), { recursive: true });
      writeFileSync(path, content, "utf8");
    }
    const result = { provider: request.provider, ...scenario.result };
    return {
      provider: request.provider,
      returnCode: scenario.returnCode,
      stdout: `${scenario.extraStdout}\n${JSON.stringify(result)}`,
      stderr: scenario.stderr,
      sessionId: `scripted-${request.sessionName ?? "session"}`,
      usage: { inputTokens: 0, outputTokens: 0 },
    };
  }
}

export function scriptedExecutorPlugin(): ExecutorPlugin {
  return {
    descriptor: {
      executorId: "scripted",
      provider: "scripted",
      runtime: "Scripted test executor",
      capabilities: new Set(["code", "filesystem", "git", "shell", "structured-output"]),
      headless: true,
      structuredOutput: true,
      persistentSessions: false,
      supportsEffort: true,
      metadata: { protocol: "scripted" },
    },
    backendFactory: () => new ScriptedExecutionBackend(),
    availabilityProbe: () => process.env.DURABLE_THREADS_ENABLE_SCRIPTED === "1",
  };
}

File: src/durable_threads/scripted.py
"""Deterministic scripted executor for offline durable-mode runs."""

from __future__ import annotations

import asyncio
import json
import os
from pathlib import Path
from typing import Any

from .execution import (
    ExecutionError,
    ExecutionOutcome,
    ExecutionRequest,
    ExecutorDescriptor,
    ExecutorPlugin,
)


class ScriptedExecutionBackend:
    """Execute a checked-in scenario from the task working directory."""

    async def execute(self, request: ExecutionRequest) -> ExecutionOutcome:
        if request.provider != "scripted":
            raise ExecutionError(
                f"backend 'scripted' cannot execute provider {request.provider!r}"
            )
        cwd = Path(request.cwd).expanduser()
        if not cwd.is_dir():
            raise ExecutionError(f"provider working directory not found: {cwd}")
        scenario_path = cwd / ".durable-threads" / "scripted.json"
        if not scenario_path.is_file():
            raise ExecutionError(f"scripted scenario not found: {scenario_path}")
        scenario = _load_scenario(scenario_path)
        delay_seconds = scenario["delaySeconds"]
        await asyncio.sleep(delay_seconds)

        for raw_path, content in scenario["writes"].items():
            path = _safe_write_path(cwd, raw_path)
            path.parent.mkdir(parents=True, exist_ok=True)
            path.write_text(content, encoding="utf-8")

        result = {"provider": request.provider, **scenario["result"]}
        stdout = f'{scenario["extraStdout"]}\n{json.dumps(result)}'
        return ExecutionOutcome(
            provider=request.provider,
            return_code=scenario["returnCode"],
            stdout=stdout,
            stderr=scenario["stderr"],
            session_id=f"scripted-{request.session_name}",
            usage={"inputTokens": 0, "outputTokens": 0},
        )


def _load_scenario(path: Path) -> dict[str, Any]:
    try:
        payload = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as exc:
        raise ExecutionError(f"could not load scripted scenario {path}: {exc}") from exc
    if not isinstance(payload, dict):
        raise ExecutionError("scripted scenario must be a JSON object")

    writes = payload.get("writes", {})
    if not isinstance(writes, dict) or any(
        not isinstance(key, str) or not isinstance(value, str)
        for key, value in writes.items()
    ):
        raise ExecutionError(
            "scripted scenario writes must be an object of string paths to strings"
        )
    result = payload.get("result")
    if not isinstance(result, dict):
        raise ExecutionError("scripted scenario result must be an object")
    return_code = payload.get("returnCode", 0)
    if type(return_code) is not int:
        raise ExecutionError("scripted scenario returnCode must be an integer")
    delay_seconds = payload.get("delaySeconds", 0.0)
    if (
        isinstance(delay_seconds, bool)
        or not isinstance(delay_seconds, (int, float))
        or delay_seconds < 0
    ):
        raise ExecutionError("scripted scenario delaySeconds must be a non-negative number")
    stderr = payload.get("stderr", "")
    if not isinstance(stderr, str):
        raise ExecutionError("scripted scenario stderr must be a string")
    extra_stdout = payload.get("extraStdout", "")
    if not isinstance(extra_stdout, str):
        raise ExecutionError("scripted scenario extraStdout must be a string")
    return {
        "writes": writes,
        "result": result,
        "returnCode": return_code,
        "delaySeconds": float(delay_seconds),
        "stderr": stderr,
        "extraStdout": extra_stdout,
    }


def _safe_write_path(cwd: Path, raw_path: str) -> Path:
    relative = Path(raw_path)
    if relative.is_absolute() or ".." in relative.parts:
        raise ExecutionError(f"scripted write path escapes cwd: {raw_path!r}")
    path = (cwd / relative).resolve()
    try:
        path.relative_to(cwd.resolve())
    except ValueError as exc:
        raise ExecutionError(f"scripted write path escapes cwd: {raw_path!r}") from exc
    return path


def scripted_executor_plugin() -> ExecutorPlugin:
    return ExecutorPlugin(
        descriptor=ExecutorDescriptor(
            executor_id="scripted",
            provider="scripted",
            runtime="Scripted test executor",
            capabilities=frozenset(
                {"code", "filesystem", "git", "shell", "structured-output"}
            ),
            headless=True,
            structured_output=True,
            persistent_sessions=False,
            supports_effort=True,
            metadata={"protocol": "scripted"},
        ),
        backend_factory=ScriptedExecutionBackend,
        availability_probe=lambda: os.environ.get("DURABLE_THREADS_ENABLE_SCRIPTED") == "1",
    )
